sys/userlib/src/arch/riscv32.rs

```diff
@@ -225,11 +225,11 @@ pub(crate) unsafe extern "C" fn sys_borrow_write_stub(
     asm!("
         # Move register arguments into place, in reverse order so that a0 is
         # loaded last when we're finished with it.
-        lw a4, 4*4(a0)
-        lw a3, 3*4(a0)
-        lw a2, 2*4(a0)
-        lw a1, 1*4(a0)
-        lw a0, 0*4(a0)
+        lw a4, 3*4(a0)
+        lw a3, 2*4(a0)
+        lw a2, 1*4(a0)
+        lw a1, 0*4(a0)
+        lw a0, 4*4(a0)
```

Comment:
Do we really need to change the assembly? I don't see the kernel side of this, but if it's just taking parameters out and putting them in memory does the order matter for the purposes of the assembly call?

Reply:
Yes, the change in assembly is required. The reason is as follows:

The structures that are being changed (*Args) are not for the user<->kernel interface. These are used within `sys/userlib` across the arch-agnostic <-> arch-dependent code boundary. Example: [`SendArgs`](https://github.com/rivosinc/hubris/blob/rivos/main-rv64/sys/userlib/src/lib.rs#L148). This is populated by [`sys_send`](https://github.com/rivosinc/hubris/blob/rivos/main-rv64/sys/userlib/src/lib.rs#L126) which is arch-agnostic code and used by [`sys_send_stub`](https://github.com/rivosinc/hubris/blob/rivos/main-rv64/sys/userlib/src/arch/riscv32.rs#L77) which is in arch-dependent code. The arch dependent code is expected to pick these parameters off the stack where the struct members are populated in registers as expected by the user<->kernel interface.

Reply:
Got it.